File: 3-Structural Patterns/3-Decorator Pattern/Ice Cream Example/Implementation with Factories/Python/ice_cream.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Ice cream module.
"""

__author__ = 'Ziang Lu'

from abc import ABC


class IceCream(ABC):
    """
    Abstract product IceCream class.
    """
    __slots__ = ['_toppings']

    def __init__(self):
        """
        Default constructor.
        """
        self._toppings = []

    def add_topping(self, topping: str) -> None:
        """
        Adds the given topping to this ice cream.
        :param topping: str
        :return: None
        """
        self._toppings.append(topping)

    def __repr__(self):
        return f'{type(self).__name__} with toppings {self._toppings}'


class VanillaIceCream(IceCream):
    """
    Concrete product VanillaIceCream class.
    """
    pass

File: 3-Structural Patterns/4-Decorator Pattern/Ice Cream Example/Implementation with Factories/Java/ice_cream/IceCream.java
package ice_cream;

import java.util.ArrayList;
import java.util.List;

/**
 * Abstract product IceCream class.
 *
 * @author Ziang Lu
 */
public abstract class IceCream {

    /**
     * Toppings of this ice cream.
     */
    private List<String> toppings = new ArrayList<>();

    /**
     * Adds the given topping to this ice cream.
     * @param topping given topping
     */
    public void addTopping(String topping) {
        toppings.add(topping);
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + " with toppings " + toppings;
    }

}
